File: SWEA/Unrated/2115. ［모의 SW 역량테스트］ 벌꿀채취/［모의 SW 역량테스트］ 벌꿀채취.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include<cstring>
using namespace std;
int T, N, M, C;
int map[11][11];

int backtrack(vector<vector<int>> & honey){
    int ret = 0;
    for(int i = 0; i < honey.size() - 1; i++){
        for(int j = i+1; j < honey.size(); j++){
            if (honey[i][0] == honey[j][0] && honey[i][1] + M > honey[j][1]) continue;
            ret = max(ret, honey[i][2] + honey[j][2]);
        }
    }
    return ret;
}

int solve(int y, int x){
    int dp[11][31];
    memset(dp, 0, sizeof(dp));
    for(int i = 0; i < M; i++){
        for(int j = 0; j <= C; j++){
            if (map[y][x+i] > j){
                dp[i+1][j] = dp[i][j];
            }
            else{
                dp[i+1][j] = max(dp[i][j], dp[i][j-map[y][x+i]] + map[y][x+i]*map[y][x+i]);
            }
        }
    }
    return dp[M][C];
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    cin >> T;
    for (int t = 0; t < T; t++){
        cin >> N >> M >> C;
        for (int y = 0; y < N; y++){
            for (int x = 0; x < N; x++){
                cin >> map[y][x];
            }
        }
        vector<vector<int>> honey;
        for (int y = 0; y < N; y++){
            for (int x = 0; x <= N - M; x++){
                honey.push_back({y, x, solve(y, x)});
            }
        }
        cout << "#" << t+1 << " " << backtrack(honey) << endl;
    }
}

File: SWEA/Unrated/2115. ［모의 SW 역량테스트］ 벌꿀채취/［모의 SW 역량테스트］ 벌꿀채취.py
def backtrack():
    res = 0
    for i in range(len(honey) - 1):
        for j in range(i + 1, len(honey)):
            if honey[i][0] == honey[j][0]:
                if honey[i][1] + M > honey[j][1]: continue
            res = max(res, honey[i][2] + honey[j][2])
    return res

def solve(y, x):
    dp = [[0] * (C + 1) for _ in range(M + 1)]
    for i in range(M):
        for j in range(C + 1):
            if arr[y][x + i] > j:
                dp[i + 1][j] = dp[i][j]
            else:
                dp[i + 1][j] = max(dp[i][j], dp[i][j - arr[y][x + i]] + arr[y][x + i] ** 2)
    return dp[M][C]

for T in range(int(input())):
    N, M, C = map(int,input().split())
    arr = [list(map(int,input().split())) for _ in range(N)]
    honey = []
    for y in range(N):
        for x in range(N - M + 1):
            honey.append([y, x, solve(y, x)])
    print(f'#{T+1}', backtrack())